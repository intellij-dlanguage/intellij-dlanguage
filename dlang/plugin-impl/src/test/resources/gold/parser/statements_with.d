unittest
{
    with (expression)
    {
        ident;
    }


    struct Foo
    {
        alias Y = int;
    }

    Y y;        // error, Y undefined
    with (Foo)
    {
        Y y;    // same as Foo.Y y;
    }

    struct S
    {
        float x;
    }

    void main()
    {
        int x;
        S s;
        with (s)
        {
            x++;  // error, shadows the int x declaration
        }
    }
}
