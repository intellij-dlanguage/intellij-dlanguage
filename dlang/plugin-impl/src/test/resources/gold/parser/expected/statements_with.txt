D Language File
  DLanguageDeclarationImpl(DECLARATION)
    UNITTEST
      UNITTEST
        PsiElement(DlangTokenType.unittest)('unittest')
        DLanguageBlockStatementImpl(BLOCK_STATEMENT)
          PsiElement(DlangTokenType.{)('{')
          DLanguageDeclarationsAndStatementsImpl(DECLARATIONS_AND_STATEMENTS)
            DLanguageDeclarationOrStatementImpl(DECLARATION_OR_STATEMENT)
              DLanguageStatementImpl(STATEMENT)
                DLanguageStatementNoCaseNoDefaultImpl(STATEMENT_NO_CASE_NO_DEFAULT)
                  DLanguageWithStatementImpl(WITH_STATEMENT)
                    DLanguageWithStatementImpl(WITH_STATEMENT)
                      PsiElement(DlangTokenType.with)('with')
                      PsiElement(DlangTokenType.()('(')
                      DLanguageExpressionImpl(EXPRESSION)
                        DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
                          DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                            DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                              DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                                DLanguageIdentifierOrTemplateInstanceImpl(IDENTIFIER_OR_TEMPLATE_INSTANCE)
                                  IDENTIFIER
                                    PsiElement(DlangTokenType.ID)('expression')
                      PsiElement(DlangTokenType.))(')')
                      DLanguageDeclarationOrStatementImpl(DECLARATION_OR_STATEMENT)
                        DLanguageStatementImpl(STATEMENT)
                          DLanguageStatementNoCaseNoDefaultImpl(STATEMENT_NO_CASE_NO_DEFAULT)
                            DLanguageBlockStatementImpl(BLOCK_STATEMENT)
                              PsiElement(DlangTokenType.{)('{')
                              DLanguageDeclarationsAndStatementsImpl(DECLARATIONS_AND_STATEMENTS)
                                DLanguageDeclarationOrStatementImpl(DECLARATION_OR_STATEMENT)
                                  DLanguageStatementImpl(STATEMENT)
                                    DLanguageStatementNoCaseNoDefaultImpl(STATEMENT_NO_CASE_NO_DEFAULT)
                                      DLanguageExpressionStatementImpl(EXPRESSION_STATEMENT)
                                        DLanguageExpressionImpl(EXPRESSION)
                                          DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
                                            DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                                              DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                                                DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                                                  DLanguageIdentifierOrTemplateInstanceImpl(IDENTIFIER_OR_TEMPLATE_INSTANCE)
                                                    IDENTIFIER
                                                      PsiElement(DlangTokenType.ID)('ident')
                                        PsiElement(DlangTokenType.;)(';')
                              PsiElement(DlangTokenType.})('}')
            DLanguageDeclarationOrStatementImpl(DECLARATION_OR_STATEMENT)
              DLanguageDeclarationImpl(DECLARATION)
                STRUCT_DECLARATION
                  PsiElement(DlangTokenType.struct)('struct')
                  IDENTIFIER
                    PsiElement(DlangTokenType.ID)('Foo')
                  DLanguageStructBodyImpl(STRUCT_BODY)
                    PsiElement(DlangTokenType.{)('{')
                    DLanguageDeclarationImpl(DECLARATION)
                      DLanguageAliasDeclarationImpl(ALIAS_DECLARATION)
                        PsiElement(DlangTokenType.alias)('alias')
                        ALIAS_INITIALIZER
                          IDENTIFIER
                            PsiElement(DlangTokenType.ID)('Y')
                          PsiElement(DlangTokenType.=)('=')
                          DLanguageTypeImpl(TYPE)
                            DLanguageType_2Impl(TYPE_2)
                              DLanguageBuiltinTypeImpl(BUILTIN_TYPE)
                                PsiElement(DlangTokenType.int)('int')
                        PsiElement(DlangTokenType.;)(';')
                    PsiElement(DlangTokenType.})('}')
            DLanguageDeclarationOrStatementImpl(DECLARATION_OR_STATEMENT)
              DLanguageDeclarationImpl(DECLARATION)
                DLanguageVariableDeclarationImpl(VARIABLE_DECLARATION)
                  DLanguageTypeImpl(TYPE)
                    DLanguageType_2Impl(TYPE_2)
                      DLanguageTypeIdentifierPartImpl(TYPE_IDENTIFIER_PART)
                        DLanguageIdentifierOrTemplateInstanceImpl(IDENTIFIER_OR_TEMPLATE_INSTANCE)
                          IDENTIFIER
                            PsiElement(DlangTokenType.ID)('Y')
                  DECLARATOR
                    IDENTIFIER
                      PsiElement(DlangTokenType.ID)('y')
                  PsiElement(DlangTokenType.;)(';')
                PsiComment(DlangTokenType.LINE_COMMENT)('// error, Y undefined')
            DLanguageDeclarationOrStatementImpl(DECLARATION_OR_STATEMENT)
              DLanguageStatementImpl(STATEMENT)
                DLanguageStatementNoCaseNoDefaultImpl(STATEMENT_NO_CASE_NO_DEFAULT)
                  DLanguageWithStatementImpl(WITH_STATEMENT)
                    DLanguageWithStatementImpl(WITH_STATEMENT)
                      PsiElement(DlangTokenType.with)('with')
                      PsiElement(DlangTokenType.()('(')
                      DLanguageExpressionImpl(EXPRESSION)
                        DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
                          DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                            DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                              DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                                DLanguageIdentifierOrTemplateInstanceImpl(IDENTIFIER_OR_TEMPLATE_INSTANCE)
                                  IDENTIFIER
                                    PsiElement(DlangTokenType.ID)('Foo')
                      PsiElement(DlangTokenType.))(')')
                      DLanguageDeclarationOrStatementImpl(DECLARATION_OR_STATEMENT)
                        DLanguageStatementImpl(STATEMENT)
                          DLanguageStatementNoCaseNoDefaultImpl(STATEMENT_NO_CASE_NO_DEFAULT)
                            DLanguageBlockStatementImpl(BLOCK_STATEMENT)
                              PsiElement(DlangTokenType.{)('{')
                              DLanguageDeclarationsAndStatementsImpl(DECLARATIONS_AND_STATEMENTS)
                                DLanguageDeclarationOrStatementImpl(DECLARATION_OR_STATEMENT)
                                  DLanguageDeclarationImpl(DECLARATION)
                                    DLanguageVariableDeclarationImpl(VARIABLE_DECLARATION)
                                      DLanguageTypeImpl(TYPE)
                                        DLanguageType_2Impl(TYPE_2)
                                          DLanguageTypeIdentifierPartImpl(TYPE_IDENTIFIER_PART)
                                            DLanguageIdentifierOrTemplateInstanceImpl(IDENTIFIER_OR_TEMPLATE_INSTANCE)
                                              IDENTIFIER
                                                PsiElement(DlangTokenType.ID)('Y')
                                      DECLARATOR
                                        IDENTIFIER
                                          PsiElement(DlangTokenType.ID)('y')
                                      PsiElement(DlangTokenType.;)(';')
                                    PsiComment(DlangTokenType.LINE_COMMENT)('// same as Foo.Y y;')
                              PsiElement(DlangTokenType.})('}')
            DLanguageDeclarationOrStatementImpl(DECLARATION_OR_STATEMENT)
              DLanguageDeclarationImpl(DECLARATION)
                STRUCT_DECLARATION
                  PsiElement(DlangTokenType.struct)('struct')
                  IDENTIFIER
                    PsiElement(DlangTokenType.ID)('S')
                  DLanguageStructBodyImpl(STRUCT_BODY)
                    PsiElement(DlangTokenType.{)('{')
                    DLanguageDeclarationImpl(DECLARATION)
                      DLanguageVariableDeclarationImpl(VARIABLE_DECLARATION)
                        DLanguageTypeImpl(TYPE)
                          DLanguageType_2Impl(TYPE_2)
                            DLanguageBuiltinTypeImpl(BUILTIN_TYPE)
                              PsiElement(DlangTokenType.float)('float')
                        DECLARATOR
                          IDENTIFIER
                            PsiElement(DlangTokenType.ID)('x')
                        PsiElement(DlangTokenType.;)(';')
                    PsiElement(DlangTokenType.})('}')
            DLanguageDeclarationOrStatementImpl(DECLARATION_OR_STATEMENT)
              DLanguageDeclarationImpl(DECLARATION)
                FUNCTION_DECLARATION
                  DLanguageTypeImpl(TYPE)
                    DLanguageType_2Impl(TYPE_2)
                      DLanguageBuiltinTypeImpl(BUILTIN_TYPE)
                        PsiElement(DlangTokenType.void)('void')
                  IDENTIFIER
                    PsiElement(DlangTokenType.ID)('main')
                  DLanguageParametersImpl(PARAMETERS)
                    PsiElement(DlangTokenType.()('(')
                    PsiElement(DlangTokenType.))(')')
                  DLanguageFunctionBodyImpl(FUNCTION_BODY)
                    DLanguageSpecifiedFunctionBodyImpl(SPECIFIED_FUNCTION_BODY)
                      DLanguageBlockStatementImpl(BLOCK_STATEMENT)
                        PsiElement(DlangTokenType.{)('{')
                        DLanguageDeclarationsAndStatementsImpl(DECLARATIONS_AND_STATEMENTS)
                          DLanguageDeclarationOrStatementImpl(DECLARATION_OR_STATEMENT)
                            DLanguageDeclarationImpl(DECLARATION)
                              DLanguageVariableDeclarationImpl(VARIABLE_DECLARATION)
                                DLanguageTypeImpl(TYPE)
                                  DLanguageType_2Impl(TYPE_2)
                                    DLanguageBuiltinTypeImpl(BUILTIN_TYPE)
                                      PsiElement(DlangTokenType.int)('int')
                                DECLARATOR
                                  IDENTIFIER
                                    PsiElement(DlangTokenType.ID)('x')
                                PsiElement(DlangTokenType.;)(';')
                          DLanguageDeclarationOrStatementImpl(DECLARATION_OR_STATEMENT)
                            DLanguageDeclarationImpl(DECLARATION)
                              DLanguageVariableDeclarationImpl(VARIABLE_DECLARATION)
                                DLanguageTypeImpl(TYPE)
                                  DLanguageType_2Impl(TYPE_2)
                                    DLanguageTypeIdentifierPartImpl(TYPE_IDENTIFIER_PART)
                                      DLanguageIdentifierOrTemplateInstanceImpl(IDENTIFIER_OR_TEMPLATE_INSTANCE)
                                        IDENTIFIER
                                          PsiElement(DlangTokenType.ID)('S')
                                DECLARATOR
                                  IDENTIFIER
                                    PsiElement(DlangTokenType.ID)('s')
                                PsiElement(DlangTokenType.;)(';')
                          DLanguageDeclarationOrStatementImpl(DECLARATION_OR_STATEMENT)
                            DLanguageStatementImpl(STATEMENT)
                              DLanguageStatementNoCaseNoDefaultImpl(STATEMENT_NO_CASE_NO_DEFAULT)
                                DLanguageWithStatementImpl(WITH_STATEMENT)
                                  DLanguageWithStatementImpl(WITH_STATEMENT)
                                    PsiElement(DlangTokenType.with)('with')
                                    PsiElement(DlangTokenType.()('(')
                                    DLanguageExpressionImpl(EXPRESSION)
                                      DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
                                        DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                                          DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                                            DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                                              DLanguageIdentifierOrTemplateInstanceImpl(IDENTIFIER_OR_TEMPLATE_INSTANCE)
                                                IDENTIFIER
                                                  PsiElement(DlangTokenType.ID)('s')
                                    PsiElement(DlangTokenType.))(')')
                                    DLanguageDeclarationOrStatementImpl(DECLARATION_OR_STATEMENT)
                                      DLanguageStatementImpl(STATEMENT)
                                        DLanguageStatementNoCaseNoDefaultImpl(STATEMENT_NO_CASE_NO_DEFAULT)
                                          DLanguageBlockStatementImpl(BLOCK_STATEMENT)
                                            PsiElement(DlangTokenType.{)('{')
                                            DLanguageDeclarationsAndStatementsImpl(DECLARATIONS_AND_STATEMENTS)
                                              DLanguageDeclarationOrStatementImpl(DECLARATION_OR_STATEMENT)
                                                DLanguageStatementImpl(STATEMENT)
                                                  DLanguageStatementNoCaseNoDefaultImpl(STATEMENT_NO_CASE_NO_DEFAULT)
                                                    DLanguageExpressionStatementImpl(EXPRESSION_STATEMENT)
                                                      DLanguageExpressionImpl(EXPRESSION)
                                                        DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
                                                          DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                                                            DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                                                              DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                                                                DLanguageIdentifierOrTemplateInstanceImpl(IDENTIFIER_OR_TEMPLATE_INSTANCE)
                                                                  IDENTIFIER
                                                                    PsiElement(DlangTokenType.ID)('x')
                                                              PsiElement(DlangTokenType.++)('++')
                                                      PsiElement(DlangTokenType.;)(';')
                                            PsiComment(DlangTokenType.LINE_COMMENT)('// error, shadows the int x declaration')
                                            PsiElement(DlangTokenType.})('}')
                        PsiElement(DlangTokenType.})('}')
          PsiElement(DlangTokenType.})('}')